File: python/day_5.py
with open('../input/day_5.txt') as f:
    input = f.read()

ordering_rules_str, orders_str = input.split('\n\n')

ordering_rules = {}
for rule in ordering_rules_str.split('\n'):
    rule = rule.split('|')
    if int(rule[1]) not in ordering_rules:
        ordering_rules[int(rule[1])] = [int(rule[0])]
    else:
        ordering_rules[int(rule[1])].append(int(rule[0]))

orders = []
for order in orders_str.split('\n'):
    if order:
        orders.append(list(map(int, order.split(','))))

part_1 = 0


def get_relevant_rules(current, order, ordering_rules):
    return list(
        filter(lambda x: x in order, ordering_rules.get(current, [])))


def is_in_order(order, ordering_rules):
    for i in range(0, len(order)):
        rules = get_relevant_rules(order[i], order, ordering_rules)
        for rule in rules:
            if rule not in order[:i]:
                return False
    return True


for order in orders:
    if is_in_order(order, ordering_rules):
        part_1 += order[len(order) // 2]

print("part 1:", part_1)


def fix(order, ordering_rules):
    for i in range(0, len(order)):
        rules = get_relevant_rules(order[i], order, ordering_rules)
        for rule in rules:
            if rule not in order[:i]:
                next = order.index(rule)
                fixed = order[:]
                fixed[next] = order[i]
                fixed[i] = order[next]
                return fixed
    return order


part_2 = 0
swaps = 0
for order in orders:
    if is_in_order(order, ordering_rules):
        continue
    fixed = order[::]
    while not is_in_order(fixed, ordering_rules):
        fixed = fix(fixed, ordering_rules)
        swaps += 1
    part_2 += fixed[len(fixed) // 2]

print("part 2:", part_2, "in", swaps, "swaps")

File: lib/Day5.hs
module Day5 (run) where

import Data.HashMap.Strict (HashMap, fromList, member, (!))
import Data.List.Split (splitOn)
import Utils (parseFile)

filePath = "input/day_5.txt"

isInOrder :: [(Int, Int)] -> [Int] -> Bool
isInOrder page_rules page_order = all (\page -> all id [elem p $ takeWhile (/= page) page_order | p <- map fst $ pagesThatMustComeBefore page_rules page_order page]) page_order

pagesThatMustComeBefore :: [(Int, Int)] -> [Int] -> Int -> [(Int, Int)]
pagesThatMustComeBefore page_rules pages page = filter (\(a, b) -> b == page && a `elem` pages) page_rules

score :: [[Int]] -> Int
score page_order = sum $ map (\pages -> pages !! (length pages `div` 2)) page_order

-- Part 1
findOrdered :: [(Int, Int)] -> [[Int]] -> Int
findOrdered page_rules page_order = score $ filter (isInOrder page_rules) page_order

-- Part 2
fixUnordered :: [(Int, Int)] -> [[Int]] -> Int
fixUnordered page_rules page_order = score $ map (fixUnordered' page_rules) $ filter (not . isInOrder page_rules) page_order
  where
    fixUnordered' :: [(Int, Int)] -> [Int] -> [Int]
    fixUnordered' page_rules page_order = if isInOrder page_rules page_order then page_order else fixUnordered' page_rules $ makeSwap page_rules page_order
    makeSwap :: [(Int, Int)] -> [Int] -> [Int]
    makeSwap page_rules page_order =
      -- TODO: find the first page that is out of order and fix that
      let (a, b) = outOfOrderPage page_rules page_order
       in swap a b page_order
    swap :: Int -> Int -> [Int] -> [Int]
    swap a b xs = map (\x -> if x == a then b else if x == b then a else x) xs

run :: IO ()
run = do
  instructionString <- parseFile filePath id
  let page_rules = map (\s -> (read $ take 2 s, read $ drop 3 s)) $ takeWhile (/= "") instructionString :: [(Int, Int)]
  let page_order = map (map read . splitOn ",") $ drop 1 $ dropWhile (/= "") instructionString :: [[Int]]
  putStr "Part 1: "
  print $ findOrdered page_rules page_order
  putStr "Part 2: "
  print $ fixUnordered page_rules page_order
